### PR TITLE
Onl 3947 toaster notifications

### PR DESCRIPTION
**Pull request by stavros-tomas**
Toasters should have a toaster component and toaster container.
The toaster component will be in chameleon.
The toaster container will be in EBO.

The toaster component should
* friendly for mobile devices / swipe
* appear on the top right

### CODE REVIEW

**Comment by matallo** [comment]
Should we add a (vertical) ellipsis? Or even force max text length.

Type prop in the Story can be a selector, better than text field.

File: src/components/ec-toaster/ec-toaster.vue [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  <div
+    v-if="messages.length > 0"
+    class="ec-toaster"
+  >
+    <transition-group
+      ref="list"
+      name="list"
+      tag="ul"
+      class="ec-toaster__list"
+      appear
```
Comment (mcibique) [inline src/components/ec-toaster/ec-toaster.vue]: A leftover from my PoC. `appear` doesn't work on transition-group. You can remove that.

File: src/components/ec-toaster/ec-toaster.vue [hunk 2]
```diff
@@ -0,0 +1,113 @@
+<template>
+  <div
+    v-if="messages.length > 0"
+    class="ec-toaster"
+  >
+    <transition-group
+      ref="list"
```
Comment (mcibique) [inline src/components/ec-toaster/ec-toaster.vue]: This ref is not used, it can be removed.

File: src/components/ec-toaster/ec-toaster.vue [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+          @change="$emit('remove', message)"
+        />
+      </li>
+    </transition-group>
+  </div>
+</template>
+
+<script>
+import EcAlert from '../ec-alert';
+import ToasterTouchDirective from '../../directives/ec-toaster-touch';
```
Comment (mcibique) [inline src/components/ec-toaster/ec-toaster.vue]: I'd like to hear more opinions here:

This directive is very toaster specific, I think we should move it to the same folder as `ec-toaster`.

File: src/components/ec-toaster/ec-toaster.vue [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+export default {
+  name: 'Toaster',
+  components: {
+    EcAlert,
+  },
+  directives: { ToasterTouch: ToasterTouchDirective },
```
Comment (mcibique) [inline src/components/ec-toaster/ec-toaster.vue]: I'd rename the import to `ToasterTouch`. Also we need to add prefix for it: `EcToasterTouch` and rename it in template too.

File: src/components/ec-toaster/ec-toaster.vue [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+      }
+    },
+  },
+};
+</script>
+
+<style lang="scss">
+.ec-toaster {
+  position: fixed;
+  z-index: 300;
```
Comment (mcibique) [inline src/components/ec-toaster/ec-toaster.vue]: Get the value from `scss/variables/`.

File: src/components/ec-toaster/ec-toaster.vue [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+  width: 100%;
+
+  &__list {
+    width: 100%;
+    display: flex;
+    flex-direction: column;
+    align-items: flex-end;
+    position: relative;
+    overflow: visible;
+    padding: 10px;
```
Comment (mcibique) [inline src/components/ec-toaster/ec-toaster.vue]: We need to add our padding: `8px` or `16px`. What's the specified value in frontify?

File: src/components/ec-toaster/ec-toaster.vue [hunk 7]
```diff
@@ -0,0 +1,113 @@
+<template>
+  <div
+    v-if="messages.length > 0"
+    class="ec-toaster"
+  >
+    <transition-group
+      ref="list"
+      name="list"
```
Comment (mcibique) [inline src/components/ec-toaster/ec-toaster.vue]: I'd rename the transition to something like `ec-toaster-items-transition`.

File: src/directives/ec-toaster-touch/ec-toaster-touch.js [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+      el.classList.add('ec-toaster__item--swipe-active');
+      oldTransform = el.style.transform;
+    };
+
+    binding.handleEnd = function handleEnd(e) {
+      const touchobj = e.changedTouches[0];
+      const distX = touchobj.pageX - startX;
+      if (distX > binding.value.minDistance) {
+        vnode.elm.dispatchEvent(new CustomEvent('toaster-touch-remove'));
```
Comment (mcibique) [inline src/directives/ec-toaster-touch/ec-toaster-touch.js]: Please remove this one as well: `ec-toaster-touch-remove`.

File: src/components/ec-toaster/ec-toaster.story.js [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+import { action } from '@storybook/addon-actions';
+import EcToaster from './ec-toaster.vue';
+
+const stories = storiesOf('Toaster', module);
+
+stories.add('basic', () => ({
+  components: { EcToaster },
+  data() {
+    return {
+      show: true,
```
Comment (mcibique) [inline src/components/ec-toaster/ec-toaster.story.js]: This data is not used in the story.

File: src/components/ec-toaster/ec-toaster.vue [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        ref="items"
+        v-toaster-touch="{ minDistance: 50 }"
+        class="ec-toaster__item"
+        @toaster-touch-remove="$emit('remove', message)"
+      >
+        <ec-alert
+          :type="message.type"
+          :title="message.title"
+          :subtitle="message.subtitle"
+          :dismissable="message.dismissable"
```
Comment (mcibique) [inline src/components/ec-toaster/ec-toaster.vue]: Can the notification be non dismissable? Like staying in the top right corner forever?

File: src/components/ec-toaster/ec-toaster.story.js [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+          type: 'success',
+          title: 'This is the title',
+          subtitle: 'This is the subtitle',
+          dismissable: true,
+        },
+        {
+          id: 2,
+          type: 'error',
+          title: 'This is the title',
+          subtitle: 'This is the subtitffffffle',
```
Comment (mcibique) [inline src/components/ec-toaster/ec-toaster.story.js]: Make our stories nicer and use lorem ipsum generator for long texts :)

File: src/components/ec-toaster/ec-toaster.story.js [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    return (
+      <div style={{
+        position: 'fixed',
+        right: 0,
+        top: 0,
+        width: '100%',
+        maxWidth: '400px',
+      }}>
+        <EcToaster messages={this.messages} onRemove={message => this.removeMessage(message)}/>
+        <button style="position:fixed;top:50%;left:50%;transform: translate(-50%, -50%);" class="ec-btn ec-btn--md ec-btn--rounded ec-btn--primary" onClick={this.addMessage}>Amend Props and Click </button>
```
Comment (mcibique) [inline src/components/ec-toaster/ec-toaster.story.js]: We should rename the button to `Add another message` or similar.

File: src/components/ec-toaster/ec-toaster.spec.js [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+    it('should render a toaster if the messages prop is passed', () => {
+      const wrapper = mountToaster({ messages });
+
+      expect(wrapper.find('.ec-toaster').exists()).toBe(true);
+      expect(wrapper.find('.ec-alert').exists()).toBe(true);
+      expect(wrapper.find('.ec-alert').classes('ec-alert--success')).toBe(true);
+      expect(wrapper.find('.ec-alert__title').text()).toBe('This is the title');
+      expect(wrapper.find('.ec-alert__subtitle').text()).toBe('This is the subtitle');
+      expect(wrapper.find('.ec-alert__dismiss-icon').exists()).toBe(true);
```
Comment (mcibique) [inline src/components/ec-toaster/ec-toaster.spec.js]: You can verify all of these in the generated snapshot. I think it is enough for the test.

File: src/components/ec-toaster/ec-toaster.spec.js [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+function mountToaster(props, mountOpts) {
+  return mount(EcToaster, {
+    propsData: { ...props },
+    ...mountOpts,
+  });
+}
+
+describe('EcToaster', () => {
+  describe('messages prop', () => {
```
Comment (mcibique) [inline src/components/ec-toaster/ec-toaster.spec.js]: I would test multiple messages with different types either.

```
let messages = ['error', 'success', ...].map((type, id) => ({ id, type, title: 'random', subtitle: 'random' }));
const wrapper = mountToaster({ messages });
expect(wrapper.element).toMatchSnapshot();
```

File: src/components/ec-toaster/ec-toaster.spec.js [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+      wrapper.find('.ec-toaster__item').trigger('touchstart', { changedTouches: [{ pageX: startX }] });
+      wrapper.find('.ec-toaster__item').trigger('touchend', { changedTouches: [{ pageX: endX }] });
+    }
+  });
+
+  it('should emit a "remove" event when dismiss button is clicked', () => {
+    const wrapper = mountToaster({ messages });
+    wrapper.find('.ec-alert__dismiss-icon').trigger('click');
+
+    expect(wrapper.emitted('remove')).toBeTruthy();
```
Comment (mcibique) [inline src/components/ec-toaster/ec-toaster.spec.js]: We should check that the `message[0]` gets passed into remove.

File: src/components/ec-toaster/ec-toaster.spec.js [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+      expect(wrapper.element).toMatchSnapshot();
+    });
+  });
+
+  describe('swipe', () => {
+    it('should emit a "remove" event when toaster is swiped for more than 50px to the right', () => {
+      const wrapper = mountToaster({ messages });
+      swipe(wrapper, 200, 251);
+
+      expect(wrapper.emitted('remove')).toBeTruthy();
```
Comment (mcibique) [inline src/components/ec-toaster/ec-toaster.spec.js]: We should check that the `message[0]` gets passed into remove.

**Comment by mcibique** [comment]
One more note: could you please show the transition to Uve and check the speed of the transition is okay? I set it to 1s in my PoC but that's because I wanted to be able to see what's happening. Maybe she would like to have them faster, e.g. 0.5s.

File: src/components/ec-toaster/ec-toaster.vue (hunk 3 above)
Comment (jaimeperezpedra) [reply src/components/ec-toaster/ec-toaster.vue]: For me it makes sense

File: src/components/ec-toaster/ec-toaster.vue (hunk 7 above)
Comment (jaimeperezpedra) [reply src/components/ec-toaster/ec-toaster.vue]: If I change it to that, the animation brakes, I cannot find why do you know how it is possible?

**Comment by jaimeperezpedra** [comment]
I spoke with Uve and she said what that animation is pretty nice, no needs more changes

File: src/components/ec-toaster/ec-toaster.vue (hunk 7 above)
Comment (mcibique) [reply src/components/ec-toaster/ec-toaster.vue]: Have you updated the classes at the bottom of the <style> block too?
```
.list-enter,
.list-leave-to {
  opacity: 0;
  transform: translateX(100%);
}
.list-leave-active {
  position: absolute;
}
```

**Pull request merged**